pkg/isb/forward/forward.go

```diff
@@ -251,14 +251,13 @@ func (isdf *InterStepDataForward) forwardAChunk(ctx context.Context) {
 
 	// forward the highest watermark to all the edges to avoid idle edge problem
 	// TODO: sort and get the highest value
-	for bufferName, offsets := range writeOffsets {
-		if publisher, ok := isdf.publishWatermark[bufferName]; ok {
+	for _, offsets := range writeOffsets {
+		for _, publisher := range isdf.publishWatermark {
```

Comment:
Is this right?

Reply:
Based on my observation, the `writeOffsets` as an edge-to-offset mapping, contains ALL the out edges and their corresponding written offsets.

In the map, some of the out edges can point to an empty offset because of falling off conditional forwarding.

Before this code change, if we wrote msgs to offset 8 for edge A and didn't write anything to edge B, we only publish offset 8 to edge A. After this change, offset 8 also gets published to edge B.

With this code change, for any written offset, we invoke ALL the publishers to publish the watermark so that ALL out edges share the same watermark.

Let me do some manual tests to double confirm.

Reply:
Any reason why we should be writing to all the edges? We should not be publishing watermarks to edges for which no data is written.

Reply:
Was trying to address https://github.com/numaproj/numaflow/blob/6870d2a4558107cf824e5ba91ff8c65125a43eed/pkg/isb/forward/forward.go#L261 , which also simplifies fetching vertex level watermark as we just need to look into one of the out edges instead of looping through all.

I need to learn more about the impact of publishing to ALL out edges. Let's sync up offline on it.